File: PY/find_leader.py
# Find the Leaders in an Array
# Question

# An element is called a leader if it is greater than all elements to its right.

# Find all leaders in the array.

# Input
# [16, 17, 4, 3, 5, 2]
# Output
# [17, 5, 2]

def find_leaders(arr):

    leaders = []
    max_right = float('-inf')

    for i in range(len(arr) - 1, -1, -1):

        if arr[i] > max_right:
            leaders.append(arr[i])
            max_right = arr[i]

    leaders.reverse()

    return leaders


arr = [16, 17, 4, 3, 5, 2]

print(find_leaders(arr))

File: JAVA/find_leader.java
import java.util.*;

public class find_leader {

    public static ArrayList<Integer> findLeaders(int[] arr) {

        ArrayList<Integer> leaders = new ArrayList<>();

        int maxRight = Integer.MIN_VALUE;

        for (int i = arr.length - 1; i >= 0; i--) {

            if (arr[i] > maxRight) {

                leaders.add(arr[i]);
                maxRight = arr[i];
            }
        }

        Collections.reverse(leaders);

        return leaders;
    }

    public static void main(String[] args) {

        int[] arr = {16, 17, 4, 3, 5, 2};

        System.out.println(findLeaders(arr));
    }
}